primitives/src/merkle_tree/field_based_mht/smt/big_lazy_merkle_tree.rs

```diff
@@ -0,0 +1,956 @@
+        let last_pos = if !self.is_tree_empty() {
+            let pos = self.get_last_non_empty_position();
+
+            // If the last non empty position is the final one of the tree, we forbid "appending"
+            if pos == self.width - 1 {
+                Err(MerkleTreeError::TooManyLeaves(self.height as usize))?
+            }
+
+            pos + 1
+        } else {
+            0
+        };
+
+        // Insert leaves inside tree
+        let mut leaves_to_insert = HashMap::new();
+        leaves_to_insert.insert(last_pos, leaf);
+        self.insert_leaves(leaves_to_insert)
+    }
+
+    fn finalize(&self) -> Result<Self, Error> {
+        let mut copy = self.clone();
+        if self.pending_changes() {
+            copy.process_leaves()?;
+        }
+        Ok(copy)
+    }
+
+    /// Finalize in place allows to continue updating the tree
+    fn finalize_in_place(&mut self) -> Result<&mut Self, Error> {
+        if self.pending_changes() {
+            self.process_leaves()?;
+        }
+        Ok(self)
+    }
+
+    fn reset(&mut self) -> &mut Self {
+        self.leaves = HashMap::new();
+        self.nodes = HashMap::new();
+        self.root = (T::ZERO_NODE_CST.unwrap().nodes[self.height as usize], false);
+        self
+    }
+
+    fn root(&self) -> Option<T::Data> {
+        if self.pending_changes() {
+            None
+        } else {
+            Some(self.root.0)
+        }
+    }
+
+    fn get_merkle_path(&self, idx: u32) -> Option<Self::MerklePath> {
+        // if root has changed then we cannot get valid merkle path until we finalize it
+        if self.pending_changes() {
+            eprintln!("Identified pending changes: unable to get path before pending changes are applied to the tree.");
+            return None;
+        }
+
+        // check that the index of the leaf is less than the width of the Merkle tree
+        if idx >= self.width {
+            eprintln!("Leaf index out of range. Max: {}, got: {}", self.width - 1, idx);
+            return None;
+        }
+
+        let mut path = Vec::with_capacity(self.height as usize);
+        let mut node_idx = idx;
+        let mut height = 0usize;
+        while height != self.height as usize {
+
+            // Estabilish if sibling is a left or right child
+            let (sibling_idx, direction) = if node_idx % T::MERKLE_ARITY as u32 == 0 {
+                (node_idx + 1, false)
+            } else {
+                (node_idx - 1, true)
+            };
+
+            // Get its hash
+            let (sibling, _) = self.get_node_at_height_and_idx(height, sibling_idx);
+
+            // Push info to path
+            path.push((sibling, direction));
+
+            height += 1; // go up one level
+            node_idx = self.width + (node_idx / T::MERKLE_ARITY as u32); // compute the index of the parent
+        }
+        debug_assert!(self.nodes.get(&node_idx).unwrap() == &self.root.0);
```

Comment:
Here I got the assertion to fail when the tree is empty.

Reply:
Thanks for reporting. There is no actual bug, just the wrong function call to get the node.
Fixed in https://github.com/HorizenOfficial/ginger-lib/pull/142/commits/e671909131ab5618f650cef53b5dbea7c44b1e86